File: set-matrix-zeroes/obzva.cpp
/**
 * 풀이
 * - matrix 전체를 탐색하다가 값이 0인 좌표를 찾으면
 *   그 좌표의 첫번째 열, 첫번째 행 좌표에 표시를 합니다
 *   if matrix[r][c] == 0
 *     matrix[r][0] = 0
 *     matrix[0][c] = 0
 * - 만약 해당 좌표가 첫번째 행이거나 첫번째 열이라면 따로 기록해둡니다
 * - 첫번째 완전탐색을 마친 후엔 matrix의 첫번째 행, 열을 보고 문제에서 요구하는 바를 수행합니다
 * 
 * Big O
 * - M: matrix의 행 개수
 * - N: matrix의 열 개수
 * 
 * - Time complexity: O(MN)
 * - Space complexity: O(1)
 *     
 */

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        bool first_row = false;
        bool first_col = false;

        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (!matrix[r][c]) {
                    if (!r) first_row = true;
                    if (!c) first_col = true;
                    matrix[r][0] = 0;
                    matrix[0][c] = 0;
                }
            }
        }

        for (int r = 1; r < m; ++r) {
            if (!matrix[r][0]) {
                for (int c = 1; c < n; ++c) matrix[r][c] = 0;
            }
        }

        for (int c = 1; c < n; ++c) {
            if (!matrix[0][c]) {
                for (int r = 1; r < m; ++r) matrix[r][c] = 0;
            }
        }

        if (first_row) {
            for (int c = 0; c < n; ++c) matrix[0][c] = 0;
        }

        if (first_col) {
            for (int r = 0; r < m; ++r) matrix[r][0] = 0;
        }
    }
};

File: rotate-image/obzva.go
/*
풀이
- matrix를 4사분면으로 나눕니다
  1사분면의 모든 좌표에 대해 아래와 같은 연산을 수행합니다
- 1사분면의 좌표 a1에 대해 a2, a3, a4를 아래처럼 정의합니다
  a2: a1을 90도 회전시켰을 때의 좌표 (2사분면에 위치함)
  a3: a2를 90도 회전시켰을 때의 좌표 (3사분면에 위치함)
  a4: a3을 90도 회전시켰을 때의 좌표 (4사분면에 위치함)
  a1 -> a2, a2 -> a3, a3 -> a4, a4 -> a1으로 값을 변경시킵니다
Big O
- N: 매트릭스의 크기
- Time complexity: O(N^2)
- Space complexity: O(1)
*/

func rotate(matrix [][]int) {
	n := len(matrix)
	// 사분면의 크기, qr, qc: 사분면의 행, 열 크기
	qr := n / 2
	qc := (n + 1) / 2

	for r := 0; r < qr; r++ {
		for c := 0; c < qc; c++ {
			r1 := r
			c1 := c

			r2 := c
			c2 := n - 1 - r

			r3 := n - 1 - r
			c3 := n - 1 - c

			r4 := n - 1 - c
			c4 := r

			matrix[r1][c1], matrix[r2][c2], matrix[r3][c3], matrix[r4][c4] = matrix[r4][c4], matrix[r1][c1], matrix[r2][c2], matrix[r3][c3]
		}
	}
}
